File: projects/plant-planner/web/garden-units.js
// garden-units.js — frontend mirror of backend `garden_units.py`. Single
// source of truth for the storage invariant introduced in migration 012:
//
//   grid_width / grid_height columns store INCHES when garden_type is one of
//     { indoor_pot, indoor_planter_box, outdoor_pot, outdoor_planter_box }
//   and FEET otherwise. pos_x / pos_y / radius_feet are ALWAYS feet.
//
// Loaded before gardens.js / render2d.js / garden.js. Keep these constants
// in lockstep with `shared-backend/routes/plant_planner/garden_units.py`.

var INCH_UNIT_GARDEN_TYPES = {
  indoor_pot: true,
  indoor_planter_box: true,
  outdoor_pot: true,
  outdoor_planter_box: true
};

var CLIMATE_CONTROLLED_GARDEN_TYPES = {
  indoor_pot: true,
  indoor_planter_box: true,
  greenhouse: true
};

function gardenTypeUsesInches(t) {
  return !!(t && INCH_UNIT_GARDEN_TYPES[t]);
}

function gardenTypeIsClimateControlled(t) {
  return !!(t && CLIMATE_CONTROLLED_GARDEN_TYPES[t]);
}

// Normalize a stored grid_width / grid_height value to feet for any garden_type.
function gridDimToFeet(value, gardenType) {
  if (value == null) return null;
  if (gardenTypeUsesInches(gardenType)) return value / 12;
  return Number(value);
}

// Display unit suffix for a garden type — "in" or "ft".
function gardenTypeUnitLabel(t) {
  return gardenTypeUsesInches(t) ? 'in' : 'ft';
}

// Pot rows store grid_width = RADIUS and grid_height = HEIGHT (vertical).
// The placement floor's planar footprint is 2r × 2r (the circle's bounding
// square) — NOT (grid_width, grid_height). Mirrors backend `floor_dims_feet`.
var POT_GARDEN_TYPES = { indoor_pot: true, outdoor_pot: true };

function gardenTypeIsPot(t) { return !!(t && POT_GARDEN_TYPES[t]); }

function floorDimsFeet(gridWidth, gridHeight, gardenType) {
  if (gardenTypeIsPot(gardenType)) {
    var radiusFt = gridDimToFeet(gridWidth, gardenType) || 0;
    var diameterFt = Math.max(0, 2 * radiusFt);
    return { width: diameterFt, length: diameterFt };
  }
  return {
    width:  gridDimToFeet(gridWidth,  gardenType) || 0,
    length: gridDimToFeet(gridHeight, gardenType) || 0
  };
}

File: shared-backend/routes/plant_planner/garden_units.py
"""Per-garden-type unit semantics. Single source of truth for the storage
invariant declared in migration 012:

    grid_width / grid_height store INCHES when garden_type is one of
    {indoor_pot, indoor_planter_box, outdoor_pot, outdoor_planter_box}
    and FEET otherwise. pos_x / pos_y / radius_feet are ALWAYS feet.

The frontend mirrors these sets in `web/garden-units.js`. Keep the two in
lockstep when adding a new garden_type.
"""

from typing import Optional

# Pots and planter boxes (indoor + outdoor) hold dimensions in inches —
# they're typically 8–48" objects measured by retailers in inches. Beds and
# greenhouses are structures measured in feet.
INCH_UNIT_TYPES = frozenset({
    "indoor_pot",
    "indoor_planter_box",
    "outdoor_pot",
    "outdoor_planter_box",
})

# Climate-controlled types skip the Location step in the wizard and default
# to indoor=True for catalog filtering. Outdoor pots are NOT in this set —
# they sit on a balcony in the user's actual hardiness zone.
CLIMATE_CONTROLLED_TYPES = frozenset({
    "indoor_pot",
    "indoor_planter_box",
    "greenhouse",
})


def garden_uses_inches(garden_type: Optional[str]) -> bool:
    return (garden_type or "") in INCH_UNIT_TYPES


def garden_is_climate_controlled(garden_type: Optional[str]) -> bool:
    return (garden_type or "") in CLIMATE_CONTROLLED_TYPES


def grid_dim_to_feet(value: Optional[int], garden_type: Optional[str]) -> Optional[float]:
    """Normalize a stored grid_width/grid_height int to feet for any garden_type."""
    if value is None:
        return None
    if garden_uses_inches(garden_type):
        return value / 12.0
    return float(value)
